**Actividad Integradora**

# 1 Importar Librería

In [112]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

import datetime


In [113]:
df = pd.read_excel('BD_Inmuebles_Limpieza_SUCIAMAS.xlsx')
df

Referencia Fecha Alta   Tipo Operación  Provincia Superficie  \
0            16 2004-01-05   Casa  Alquiler     Girona        179   
1            17 2004-01-05   Casa  Alquiler  Tarragona         58   
2            23 2004-01-10   Casa  Alquiler     Girona        183   
3            28 2004-01-12   Casa  Alquiler  Barcelona        187   
4            29 2004-01-12   Casa  Alquiler  Tarragona         91   
..          ...        ...    ...       ...        ...        ...   
495         452 2004-06-19  Suelo     Venta     Lleida        215   
496         453 2004-06-19  Suelo     Venta     Girona        185   
497         472 2004-06-25  Suelo     Venta  Tarragona        297   
498         474 2004-06-25  Suelo     Venta     Girona        199   
499         480 2004-06-28  Suelo     Venta     Lleida         58   

    Precio Venta Fecha Venta Vendedor Edad del vendedor  Numero cuartos  
0         937960  2004-11-21   Carmen                43             1.0  
1         358846  2004-10-08    Luisa                22             2.0  
2        1438929  2004-04-21    Luisa                70             1.0  
3        1660560  2004-06-16  Joaquín                55             NaN  
4         753571  2004-07-07    Luisa                26             2.0  
..           ...         ...      ...               ...             ...  
495      2097970  2004-11-13    Pedro                45             NaN  
496      1616900  2005-04-08    Jesús                35             NaN  
497      1820016  2005-01-08    María               NaN             2.0  
498      1622646  2005-06-23    Jesús               NaN             NaN  
499       530816  2004-10-07    Luisa                55             2.0  

[500 rows x 11 columns]

In [114]:
#Reiniciamos el index
df.reset_index()

index  Referencia Fecha Alta   Tipo Operación  Provincia Superficie  \
0        0          16 2004-01-05   Casa  Alquiler     Girona        179   
1        1          17 2004-01-05   Casa  Alquiler  Tarragona         58   
2        2          23 2004-01-10   Casa  Alquiler     Girona        183   
3        3          28 2004-01-12   Casa  Alquiler  Barcelona        187   
4        4          29 2004-01-12   Casa  Alquiler  Tarragona         91   
..     ...         ...        ...    ...       ...        ...        ...   
495    495         452 2004-06-19  Suelo     Venta     Lleida        215   
496    496         453 2004-06-19  Suelo     Venta     Girona        185   
497    497         472 2004-06-25  Suelo     Venta  Tarragona        297   
498    498         474 2004-06-25  Suelo     Venta     Girona        199   
499    499         480 2004-06-28  Suelo     Venta     Lleida         58   

    Precio Venta Fecha Venta Vendedor Edad del vendedor  Numero cuartos  
0         937960  2004-11-21   Carmen                43             1.0  
1         358846  2004-10-08    Luisa                22             2.0  
2        1438929  2004-04-21    Luisa                70             1.0  
3        1660560  2004-06-16  Joaquín                55             NaN  
4         753571  2004-07-07    Luisa                26             2.0  
..           ...         ...      ...               ...             ...  
495      2097970  2004-11-13    Pedro                45             NaN  
496      1616900  2005-04-08    Jesús                35             NaN  
497      1820016  2005-01-08    María               NaN             2.0  
498      1622646  2005-06-23    Jesús               NaN             NaN  
499       530816  2004-10-07    Luisa                55             2.0  

[500 rows x 12 columns]

In [115]:
#Cambiamos el nombre del index por Id
df.index.names = ['Id']

#Imprimimos el dataframe
df.head(10)

Referencia Fecha Alta  Tipo Operación  Provincia Superficie Precio Venta  \
Id                                                                             
0           16 2004-01-05  Casa  Alquiler     Girona        179       937960   
1           17 2004-01-05  Casa  Alquiler  Tarragona         58       358846   
2           23 2004-01-10  Casa  Alquiler     Girona        183      1438929   
3           28 2004-01-12  Casa  Alquiler  Barcelona        187      1660560   
4           29 2004-01-12  Casa  Alquiler  Tarragona         91       753571   
5           32 2004-01-13  Casa  Alquiler     Girona         92       627348   
6           70 2004-01-28  Casa  Alquiler        NaN        179      1647695   
7          103 2004-02-08  Casa  Alquiler     Girona        258      1188090   
8          122 2004-02-17  Casa  Alquiler  Barcelona        270      1385910   
9          150 2004-03-01  Casa  Alquiler     Lleida        212      1800516   

   Fecha Venta Vendedor Edad del vendedor  Numero cuartos  
Id                                                         
0   2004-11-21   Carmen                43             1.0  
1   2004-10-08    Luisa                22             2.0  
2   2004-04-21    Luisa                70             1.0  
3   2004-06-16  Joaquín                55             NaN  
4   2004-07-07    Luisa                26             2.0  
5   2004-04-12    María                12             2.0  
6   2004-07-12    María                44             NaN  
7   2004-07-19    María                35             2.0  
8   2005-02-05    Pedro                34             2.0  
9   2004-12-19    Jesús                22             2.0

# 3 Análisis exploratorio de datos (EDA)

Proceso:
*   Forma de la base de datos
*   Ver la cabeza y la cola
*   Tipos de datos
*   Estadística básica

La exploración de análisis también utiliza gráficas, por lo que este proceso lo haremos también después de la limpieza para comparar.






## Forma de la base de datos

In [116]:
#Usamos función shape para el total de renglones y columnas.
df.shape

(500, 11)

In [117]:
#Podemos ver solamente los renglones
len(df) #Renglones

500

In [118]:
#Podemos ver solamente las columnas
len(df.columns) #Columnas

11

## Ver los primeros (head) y los últimos (tail) registros
Este punto lo realizamos para una exploración rápida de los datos

In [119]:
# Primeros 5 renglones
df.head()

Referencia Fecha Alta  Tipo Operación  Provincia Superficie Precio Venta  \
Id                                                                             
0           16 2004-01-05  Casa  Alquiler     Girona        179       937960   
1           17 2004-01-05  Casa  Alquiler  Tarragona         58       358846   
2           23 2004-01-10  Casa  Alquiler     Girona        183      1438929   
3           28 2004-01-12  Casa  Alquiler  Barcelona        187      1660560   
4           29 2004-01-12  Casa  Alquiler  Tarragona         91       753571   

   Fecha Venta Vendedor Edad del vendedor  Numero cuartos  
Id                                                         
0   2004-11-21   Carmen                43             1.0  
1   2004-10-08    Luisa                22             2.0  
2   2004-04-21    Luisa                70             1.0  
3   2004-06-16  Joaquín                55             NaN  
4   2004-07-07    Luisa                26             2.0

In [120]:
# Últimos 5 renglones
df.tail()

Referencia Fecha Alta   Tipo Operación  Provincia Superficie  \
Id                                                                  
495         452 2004-06-19  Suelo     Venta     Lleida        215   
496         453 2004-06-19  Suelo     Venta     Girona        185   
497         472 2004-06-25  Suelo     Venta  Tarragona        297   
498         474 2004-06-25  Suelo     Venta     Girona        199   
499         480 2004-06-28  Suelo     Venta     Lleida         58   

    Precio Venta Fecha Venta Vendedor Edad del vendedor  Numero cuartos  
Id                                                                       
495      2097970  2004-11-13    Pedro                45             NaN  
496      1616900  2005-04-08    Jesús                35             NaN  
497      1820016  2005-01-08    María               NaN             2.0  
498      1622646  2005-06-23    Jesús               NaN             NaN  
499       530816  2004-10-07    Luisa                55             2.0

## Tipo de datos

Vemos que el tipo de datos sea correcto

In [121]:
df.dtypes

Referencia                    int64
Fecha Alta           datetime64[ns]
Tipo                         object
Operación                    object
Provincia                    object
Superficie                   object
Precio Venta                 object
Fecha Venta          datetime64[ns]
Vendedor                     object
Edad del vendedor            object
Numero cuartos              float64
dtype: object

In [122]:
#Vemos las columnas y su tipo correspondiente, en donde nos dice si se contiene nulos o no
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 500 entries, 0 to 499
Data columns (total 11 columns):
 #   Column             Non-Null Count  Dtype         
---  ------             --------------  -----         
 0   Referencia         500 non-null    int64         
 1   Fecha Alta         500 non-null    datetime64[ns]
 2   Tipo               498 non-null    object        
 3   Operación          499 non-null    object        
 4   Provincia          492 non-null    object        
 5   Superficie         500 non-null    object        
 6   Precio Venta       500 non-null    object        
 7   Fecha Venta        500 non-null    datetime64[ns]
 8   Vendedor           499 non-null    object        
 9   Edad del vendedor  447 non-null    object        
 10  Numero cuartos     234 non-null    float64       
dtypes: datetime64[ns](2), float64(1), int64(1), object(7)
memory usage: 43.1+ KB


Como se puede observar, las columnas Superficie, Precio_venta y Edad_del_vendedor están marcados como objects cuando claramente son numéricos.

También se observa que tenemos columnas con valores nulos en Tipo, Operación, Provincia, Vendedor, Edad del vendedor y Numero cuartos

## Estadística básica

Obtenemos la estadística básica para ver la distribución de los datos

In [123]:
#Podemos ver la distribución de los datos
df.describe()

Referencia                  Fecha Alta                 Fecha Venta  \
count  500.000000                         500                         500   
mean   250.500000  2004-04-05 19:23:31.200000  2004-11-22 00:51:50.400000   
min      1.000000         2004-01-01 00:00:00         2004-04-05 00:00:00   
25%    125.750000         2004-02-19 00:00:00         2004-09-13 00:00:00   
50%    250.500000         2004-04-09 00:00:00         2004-11-17 00:00:00   
75%    375.250000         2004-05-22 06:00:00         2005-02-01 00:00:00   
max    500.000000         2004-07-04 00:00:00         2005-07-03 00:00:00   
std    144.481833                         NaN                         NaN   

       Numero cuartos  
count      234.000000  
mean         2.854701  
min          1.000000  
25%          2.000000  
50%          3.000000  
75%          4.000000  
max          6.000000  
std          1.483905

## Identificar nulos

In [124]:
#El total de nulos por columnas
total_of_all = df.isnull().sum()
total_of_all


Referencia             0
Fecha Alta             0
Tipo                   2
Operación              1
Provincia              8
Superficie             0
Precio Venta           0
Fecha Venta            0
Vendedor               1
Edad del vendedor     53
Numero cuartos       266
dtype: int64

In [125]:
#Total de porcentaje de nulos por columnas
percent_of_all = df.isnull().mean() # regresa la fracción de valores nulos
missing_data_test = pd.concat([total_of_all, percent_of_all*100], axis=1, keys=['Total Nulos', 'Percent (%)'])
missing_data_test

Total Nulos  Percent (%)
Referencia                   0          0.0
Fecha Alta                   0          0.0
Tipo                         2          0.4
Operación                    1          0.2
Provincia                    8          1.6
Superficie                   0          0.0
Precio Venta                 0          0.0
Fecha Venta                  0          0.0
Vendedor                     1          0.2
Edad del vendedor           53         10.6
Numero cuartos             266         53.2

## Adicionalmente usamos unique para ver los valores únicos de las columnas de interés

### Columna Tipo

In [126]:
df.Tipo.unique()

array(['Casa', 'Industrial', 'Local', nan, 'Oficina', 'Parking', 'Piso',
       'Pisar', 'Suelo'], dtype=object)

In [127]:
#Nos da el conteo de cada valor en orden descentente.
df.Tipo.value_counts()

#Otra manera de escribirlo
#df['Tipo'].value_counts()


Tipo
Suelo         82
Oficina       75
Industrial    74
Parking       74
Local         73
Casa          57
Piso          50
Pisar         13
Name: count, dtype: int64

La columna Tipo tiene el valor de "Pisar" que debe de cambiarse a Piso


### Columna Operación

In [128]:
df.Operación.unique()

array(['Alquiler', 'Venta', 'Alkiler', ' ', nan], dtype=object)

In [129]:
#Nos da el conteo de cada valor en orden descendente.
df.Operación.value_counts()

#Otra manera de escribirlo
#df['Operación'].value_counts()


Operación
Alquiler    249
Venta       243
Alkiler       6
              1
Name: count, dtype: int64

La columna Operación tiene escrita mal la palabra "Alkiler" y debe de cambiarse a Alquiler.

Tiene un valor en espacio.

### Columna Provincia

In [130]:
df.Provincia.unique()

array(['Girona', 'Tarragona', 'Barcelona', nan, 'Lleida', ' '],
      dtype=object)

In [131]:
#Nos da el conteo de cada valor en orden descentente.
df.Provincia.value_counts()

#Otra manera de escribirlo
#df['Provincia'].value_counts()


Provincia
Girona       135
Barcelona    122
Tarragona    121
Lleida       113
               1
Name: count, dtype: int64

La columna Provincia tiene un valor con espacio

### Columna Superficie

In [132]:
df.Superficie.unique()

array([179, 58, 183, 187, 91, 92, 258, 270, 212, 100, 87, 63, 56, 170,
       201, 181, 177, 295, 178, 298, 251, 149, 192, 72, 57, 143, 65, 281,
       278, 120, 235, 227, 130, 119, 284, 211, 156, 239, 204, 285, 229,
       146, 241, 141, 80, 123, 173, 90, 214, 155, 153, 261, 60, 61, 131,
       176, 142, 257, 140, 219, 112, 236, 93, 132, 103, 52, ' ', 193, 216,
       46, 209, 159, 150, 225, 277, 190, 242, 76, 247, 74, 287, 233, 248,
       268, 240, 228, 166, 54, 118, 194, 139, 88, 265, 163, 263, 226, 223,
       238, 106, 45, 196, 121, 189, 162, 243, 70, 206, 274, 68, 127, 49,
       125, 160, 231, 117, 169, 82, 182, 246, 237, 138, 259, 137, 203,
       260, 0, 44, 199, 40, 55, 109, 202, 291, 267, 230, 269, 224, 148,
       126, 215, 282, 79, 94, 124, 116, 275, 105, 134, 128, 174, 186, 244,
       110, 41, 207, 205, 77, 168, 144, 104, 253, 172, 42, 154, 86, 133,
       249, 102, 158, 107, 232, 69, 62, 217, 95, 152, 184, 296, 271, 234,
       299, 300, 51, 167, 164, 175, 200, 197, 20

In [133]:
#Nos da el conteo de cada valor en orden descentente.
df.Superficie.value_counts()

#Otra manera de escribirlo
#df['Superficie'].value_counts()


Superficie
       10
179     5
298     5
209     5
54      5
       ..
267     1
109     1
40      1
44      1
297     1
Name: count, Length: 228, dtype: int64

La columna Superficie tiene 10 valores con espacios

### Columna Precio Venta

In [134]:
df["Precio Venta"].unique()

array([937960, 358846, 1438929, 1660560, 753571, 627348, 1647695, 1188090,
       1385910, 1800516, 731700, 779868, 492156, 474600, 995520, 1107108,
       1449629, 924294, 1877675, 1458176, 2612566, 2439971, 1238339,
       1159296, 702576, 419178, 836836, 546780, 2483197, 1537062, 1140840,
       '.', 1467782, 1230190, 686154, 1561432, 875228, 537264, 1369680,
       1670965, 1176597, 1291116, 2534505, 1990697, 1083466, 2060550,
       732495, 354320, 775884, 1117753, 412740, 1962380, '$', 1506438,
       2475063, 253920, 531432, 953156, 820336, 1405374, 1799771, 1054340,
       948927, 1104992, 1812244, 778875, 723609, 922020, 965358, 728931,
       2067236, 508924, '%', 1658835, 1398600, 351256, 1042283, 1582527,
       968250, 712712, 1360800, 1171894, 2571111, 2456159, 977740,
       1227224, 2110966, 1263495, 501220, 1210053, 1331166, 2188138,
       729788, 2452989, 1541528, 1674744, 2628812, 2238000, 999328,
       2020992, 1457488, '¡', 1373816, 1142938, 509010, 1098055, 4387

In [135]:
#Nos da el conteo de cada valor en orden descentente.
df["Precio Venta"].value_counts()

#Otra manera de escribirlo
#df['Precio_venta'].value_counts()


Precio Venta
0          3
+          2
           2
$          2
*          2
          ..
1992228    1
1394461    1
877680     1
391391     1
530816     1
Name: count, Length: 494, dtype: int64

La columna Precio_venta tiene valores vacíos, ceros y múltiples simbolos que deben de ser corregidos

### Columna Vendedor

In [136]:
df.Vendedor.unique()

array(['Carmen', 'Luisa', 'Joaquín', 'María', 'Pedro', 'Jesús', nan],
      dtype=object)

In [137]:
#Nos da el conteo de cada valor en orden descentente.
df.Vendedor.value_counts()

#Otra manera de escribirlo
#df['Vendedor'].value_counts()


Vendedor
Luisa      100
Jesús       95
Pedro       93
Carmen      82
María       67
Joaquín     62
Name: count, dtype: int64

La columna vendedor tiene valores nulos

### Columna Edad del vendedor

In [138]:
df["Edad del vendedor"].unique()

array([43, 22, 70, 55, 26, 12, 44, 35, 34, 45, 66, nan, 56, 87, 57, 38,
       76, 20, 88, 224, 65, 27, 78, 46, ' ', 74, 33, 54, 67, 24, 77, 89,
       188], dtype=object)

In [139]:
#Nos da el conteo de cada valor en orden descentente.
df["Edad del vendedor"].value_counts()

#Otra manera de escribirlo
#df["Edad del vendedor"].value_counts()


Edad del vendedor
44     50
43     39
55     37
56     30
45     28
88     28
22     27
35     26
78     26
34     16
26     14
38     14
27     13
70     13
33     12
54      8
66      8
46      8
76      7
224     6
65      4
57      4
74      4
87      4
67      4
24      4
77      4
        3
12      3
89      1
20      1
188     1
Name: count, dtype: int64

La columna edad_vendedor tiene campos vacíos y edades que no entran dentro de un límite lógico como 224 y 12 años

### Columna Numero cuartos

In [140]:
df["Numero cuartos"].unique()

array([ 1.,  2., nan,  3.,  5.,  6.,  4.])

In [141]:
#Nos da el conteo de cada valor en orden descentente.
df["Numero cuartos"].value_counts()

#Otra manera de escribirlo
#df["Numero cuartos"].value_counts()


Numero cuartos
3.0    65
1.0    55
2.0    45
4.0    31
5.0    24
6.0    14
Name: count, dtype: int64

La columna Numero_cuartos tiene muchos campos vacíos

# 4 Preprocesamiento de los datos

Proceso:
*   Renombrar columnas (de ser necesario)
*   Eliminar columnas que no utilicemos
*   Eliminar las columnas que tengan alto porcentaje de valores nulos
*   Eliminar duplicados
*   Corregir el formato del tipo de datos
*   Obten una muestra de la base de datos (en caso de que la base de datos sea muy grande)
*   Elimina outliers y valores vacíos
*   Eliminar duplicados
*   Imputación de datos





## Renombrar columnas (de ser necesario)

In [142]:
df =df.set_axis(['Referencia','Fecha_alta','Tipo','Operación','Provincia',"Superficie", "Precio_venta", "Fecha_venta", "Vendedor", "Edad_vendedor","Numero_cuartos"], axis=1)
df
#Ahora se muestra el nuevo Data frame con los nombres por columna actualizados.

Referencia Fecha_alta   Tipo Operación  Provincia Superficie  \
Id                                                                  
0            16 2004-01-05   Casa  Alquiler     Girona        179   
1            17 2004-01-05   Casa  Alquiler  Tarragona         58   
2            23 2004-01-10   Casa  Alquiler     Girona        183   
3            28 2004-01-12   Casa  Alquiler  Barcelona        187   
4            29 2004-01-12   Casa  Alquiler  Tarragona         91   
..          ...        ...    ...       ...        ...        ...   
495         452 2004-06-19  Suelo     Venta     Lleida        215   
496         453 2004-06-19  Suelo     Venta     Girona        185   
497         472 2004-06-25  Suelo     Venta  Tarragona        297   
498         474 2004-06-25  Suelo     Venta     Girona        199   
499         480 2004-06-28  Suelo     Venta     Lleida         58   

    Precio_venta Fecha_venta Vendedor Edad_vendedor  Numero_cuartos  
Id                                                                   
0         937960  2004-11-21   Carmen            43             1.0  
1         358846  2004-10-08    Luisa            22             2.0  
2        1438929  2004-04-21    Luisa            70             1.0  
3        1660560  2004-06-16  Joaquín            55             NaN  
4         753571  2004-07-07    Luisa            26             2.0  
..           ...         ...      ...           ...             ...  
495      2097970  2004-11-13    Pedro            45             NaN  
496      1616900  2005-04-08    Jesús            35             NaN  
497      1820016  2005-01-08    María           NaN             2.0  
498      1622646  2005-06-23    Jesús           NaN             NaN  
499       530816  2004-10-07    Luisa            55             2.0  

[500 rows x 11 columns]

## Eliminar columnas que no utilicemos



In [143]:
df=df.drop(df.columns[[0,1]], axis = 1)
df.head()

Tipo Operación  Provincia Superficie Precio_venta Fecha_venta Vendedor  \
Id                                                                           
0   Casa  Alquiler     Girona        179       937960  2004-11-21   Carmen   
1   Casa  Alquiler  Tarragona         58       358846  2004-10-08    Luisa   
2   Casa  Alquiler     Girona        183      1438929  2004-04-21    Luisa   
3   Casa  Alquiler  Barcelona        187      1660560  2004-06-16  Joaquín   
4   Casa  Alquiler  Tarragona         91       753571  2004-07-07    Luisa   

   Edad_vendedor  Numero_cuartos  
Id                                
0             43             1.0  
1             22             2.0  
2             70             1.0  
3             55             NaN  
4             26             2.0

In [144]:
#Podemos usar del para eliminar columnas
#del df["Vendedor"]
df

Tipo Operación  Provincia Superficie Precio_venta Fecha_venta Vendedor  \
Id                                                                             
0     Casa  Alquiler     Girona        179       937960  2004-11-21   Carmen   
1     Casa  Alquiler  Tarragona         58       358846  2004-10-08    Luisa   
2     Casa  Alquiler     Girona        183      1438929  2004-04-21    Luisa   
3     Casa  Alquiler  Barcelona        187      1660560  2004-06-16  Joaquín   
4     Casa  Alquiler  Tarragona         91       753571  2004-07-07    Luisa   
..     ...       ...        ...        ...          ...         ...      ...   
495  Suelo     Venta     Lleida        215      2097970  2004-11-13    Pedro   
496  Suelo     Venta     Girona        185      1616900  2005-04-08    Jesús   
497  Suelo     Venta  Tarragona        297      1820016  2005-01-08    María   
498  Suelo     Venta     Girona        199      1622646  2005-06-23    Jesús   
499  Suelo     Venta     Lleida         58       530816  2004-10-07    Luisa   

    Edad_vendedor  Numero_cuartos  
Id                                 
0              43             1.0  
1              22             2.0  
2              70             1.0  
3              55             NaN  
4              26             2.0  
..            ...             ...  
495            45             NaN  
496            35             NaN  
497           NaN             2.0  
498           NaN             NaN  
499            55             2.0  

[500 rows x 9 columns]

## Eliminar las columnas que tengan alto porcentaje de valores nulos

El porcentaje que decidmos depende del contexto de nuestro problema.


In [145]:
#Ej. Borrar columnas con el 40 a 100% de nulos
df.head()

Tipo Operación  Provincia Superficie Precio_venta Fecha_venta Vendedor  \
Id                                                                           
0   Casa  Alquiler     Girona        179       937960  2004-11-21   Carmen   
1   Casa  Alquiler  Tarragona         58       358846  2004-10-08    Luisa   
2   Casa  Alquiler     Girona        183      1438929  2004-04-21    Luisa   
3   Casa  Alquiler  Barcelona        187      1660560  2004-06-16  Joaquín   
4   Casa  Alquiler  Tarragona         91       753571  2004-07-07    Luisa   

   Edad_vendedor  Numero_cuartos  
Id                                
0             43             1.0  
1             22             2.0  
2             70             1.0  
3             55             NaN  
4             26             2.0

In [146]:
#Eliminar columnas con el nombre
df.drop('Numero_cuartos', inplace=True, axis=1)

#Otra forma de eliminar columnas
#df.drop(columns='Numero_cuartos')

df

Tipo Operación  Provincia Superficie Precio_venta Fecha_venta Vendedor  \
Id                                                                             
0     Casa  Alquiler     Girona        179       937960  2004-11-21   Carmen   
1     Casa  Alquiler  Tarragona         58       358846  2004-10-08    Luisa   
2     Casa  Alquiler     Girona        183      1438929  2004-04-21    Luisa   
3     Casa  Alquiler  Barcelona        187      1660560  2004-06-16  Joaquín   
4     Casa  Alquiler  Tarragona         91       753571  2004-07-07    Luisa   
..     ...       ...        ...        ...          ...         ...      ...   
495  Suelo     Venta     Lleida        215      2097970  2004-11-13    Pedro   
496  Suelo     Venta     Girona        185      1616900  2005-04-08    Jesús   
497  Suelo     Venta  Tarragona        297      1820016  2005-01-08    María   
498  Suelo     Venta     Girona        199      1622646  2005-06-23    Jesús   
499  Suelo     Venta     Lleida         58       530816  2004-10-07    Luisa   

    Edad_vendedor  
Id                 
0              43  
1              22  
2              70  
3              55  
4              26  
..            ...  
495            45  
496            35  
497           NaN  
498           NaN  
499            55  

[500 rows x 8 columns]

## Eliminar duplicados

In [147]:
#Vemos los duplicados
df.duplicated()

Id
0      False
1      False
2      False
3      False
4      False
       ...  
495    False
496    False
497    False
498    False
499    False
Length: 500, dtype: bool

In [148]:
#Eliminar duplicados completa
df = df.drop_duplicates()
df

Tipo Operación  Provincia Superficie Precio_venta Fecha_venta Vendedor  \
Id                                                                             
0     Casa  Alquiler     Girona        179       937960  2004-11-21   Carmen   
1     Casa  Alquiler  Tarragona         58       358846  2004-10-08    Luisa   
2     Casa  Alquiler     Girona        183      1438929  2004-04-21    Luisa   
3     Casa  Alquiler  Barcelona        187      1660560  2004-06-16  Joaquín   
4     Casa  Alquiler  Tarragona         91       753571  2004-07-07    Luisa   
..     ...       ...        ...        ...          ...         ...      ...   
495  Suelo     Venta     Lleida        215      2097970  2004-11-13    Pedro   
496  Suelo     Venta     Girona        185      1616900  2005-04-08    Jesús   
497  Suelo     Venta  Tarragona        297      1820016  2005-01-08    María   
498  Suelo     Venta     Girona        199      1622646  2005-06-23    Jesús   
499  Suelo     Venta     Lleida         58       530816  2004-10-07    Luisa   

    Edad_vendedor  
Id                 
0              43  
1              22  
2              70  
3              55  
4              26  
..            ...  
495            45  
496            35  
497           NaN  
498           NaN  
499            55  

[500 rows x 8 columns]

In [149]:
#Eliminación de los registros duplicados parciales
df.duplicated(df.columns[~df.columns.isin(['Vendedor'])])

Id
0      False
1      False
2      False
3      False
4      False
       ...  
495    False
496    False
497    False
498    False
499    False
Length: 500, dtype: bool

## Obten una muestra de la base de datos (en caso de que la base de datos sea muy grande)

In [150]:
# Obten una muestra para trabajar de forma aleatoria

df.sample() #Nos da un solo renglón
df.sample(400) # Puedes usar directamente el número
df_sample=df.sample(n=400)
df_sample

Tipo Operación  Provincia Superficie Precio_venta Fecha_venta  \
Id                                                                         
160       Local  Alquiler     Girona        138       877680  2005-03-31   
256     Oficina     Venta  Tarragona         42       367584  2004-08-11   
155       Local  Alquiler  Tarragona        284      1431076  2004-12-22   
473       Suelo     Venta  Tarragona        148      1431012  2004-09-01   
441       Suelo  Alquiler     Girona        262      2404898  2004-10-20   
..          ...       ...        ...        ...          ...         ...   
248     Oficina     Venta     Girona        253      2047276  2004-09-10   
465       Suelo     Venta     Lleida         56       471072  2004-05-29   
339     Parking     Venta  Tarragona        234      1398852  2004-11-01   
442       Suelo  Alquiler     Girona        288      2699136  2004-07-28   
63   Industrial  Alquiler  Tarragona        112      1104992  2004-06-25   

    Vendedor Edad_vendedor  
Id                          
160    Luisa            44  
256  Joaquín            44  
155    María            22  
473    Jesús            38  
441    Luisa            27  
..       ...           ...  
248    Luisa            74  
465    María            43  
339    María            26  
442    Pedro           188  
63     Pedro            45  

[400 rows x 8 columns]

In [151]:
#Para generar una submuestra de tamaño mayor usamos el parametro replace = True
df.sample(n=5000, replace=True)


Tipo Operación  Provincia Superficie Precio_venta Fecha_venta  \
Id                                                                         
221     Oficina  Alquiler     Girona        174       720882  2004-08-14   
80   Industrial  Alquiler  Barcelona        225      1360800  2004-07-22   
475       Suelo     Venta  Tarragona         59       453651  2004-10-07   
314     Parking  Alquiler     Lleida        299      1948583  2004-12-21   
171       Local     Venta  Tarragona         68       664700  2004-10-01   
..          ...       ...        ...        ...          ...         ...   
219     Oficina  Alquiler     Lleida        128      1259904  2004-10-02   
83   Industrial  Alquiler     Girona        277      2456159  2004-08-04   
66   Industrial  Alquiler  Barcelona        123       723609  2004-12-16   
202       Local     Venta  Tarragona        215       870105  2005-04-02   
107  Industrial     Venta     Girona        118      1067192  2004-08-25   

    Vendedor Edad_vendedor  
Id                          
221    Pedro            35  
80     Jesús                
475    María            44  
314    María            12  
171      NaN            78  
..       ...           ...  
219    Luisa            88  
83     María            33  
66   Joaquín                
202    Jesús           NaN  
107   Carmen            34  

[5000 rows x 8 columns]

In [152]:
#Un porcentaje del total
df.sample(frac=0)

df.sample(frac=0.5).shape

df.sample(frac=1).shape

(500, 8)

In [153]:
# Primeros 1000 Renglones - Podemos usarlos para explorar las columnas y corregir el tipo de datos
df.head(300).to_csv('Actividad.csv')

In [154]:
# Ultimos 1000 Renglones - Podemos usarlos para explorar las columnas y corregir el tipo de datos
df.tail(300).to_csv('Actividad2.csv')

## Corregir el formato del tipo de datos


In [155]:
df.isnull().sum()

Tipo              2
Operación         1
Provincia         8
Superficie        0
Precio_venta      0
Fecha_venta       0
Vendedor          1
Edad_vendedor    53
dtype: int64

In [156]:
df.dtypes

Tipo                     object
Operación                object
Provincia                object
Superficie               object
Precio_venta             object
Fecha_venta      datetime64[ns]
Vendedor                 object
Edad_vendedor            object
dtype: object

### Cambiamos signos especiales usando pd.to_numeric coerce

Coerce convierte los valores no númericos a NaN



In [157]:
# Hacemos una copia para no modificar el dataframe original
df_coerce = df.copy()

# Usando pd.to_numeric, se cambian los errores a nulos
df_coerce['Superficie'] = pd.to_numeric(df_coerce['Superficie'], errors= 'coerce')
df_coerce['Precio_venta'] = pd.to_numeric(df_coerce['Precio_venta'], errors= 'coerce')
df_coerce['Edad_vendedor'] = pd.to_numeric(df_coerce['Edad_vendedor'], errors= 'coerce') #.isnull() nos va a decir si es verdadero o falso
df_coerce

Tipo Operación  Provincia  Superficie  Precio_venta Fecha_venta  \
Id                                                                      
0     Casa  Alquiler     Girona       179.0      937960.0  2004-11-21   
1     Casa  Alquiler  Tarragona        58.0      358846.0  2004-10-08   
2     Casa  Alquiler     Girona       183.0     1438929.0  2004-04-21   
3     Casa  Alquiler  Barcelona       187.0     1660560.0  2004-06-16   
4     Casa  Alquiler  Tarragona        91.0      753571.0  2004-07-07   
..     ...       ...        ...         ...           ...         ...   
495  Suelo     Venta     Lleida       215.0     2097970.0  2004-11-13   
496  Suelo     Venta     Girona       185.0     1616900.0  2005-04-08   
497  Suelo     Venta  Tarragona       297.0     1820016.0  2005-01-08   
498  Suelo     Venta     Girona       199.0     1622646.0  2005-06-23   
499  Suelo     Venta     Lleida        58.0      530816.0  2004-10-07   

    Vendedor  Edad_vendedor  
Id                           
0     Carmen           43.0  
1      Luisa           22.0  
2      Luisa           70.0  
3    Joaquín           55.0  
4      Luisa           26.0  
..       ...            ...  
495    Pedro           45.0  
496    Jesús           35.0  
497    María            NaN  
498    Jesús            NaN  
499    Luisa           55.0  

[500 rows x 8 columns]

In [158]:
df_coerce.isnull().sum()

Tipo              2
Operación         1
Provincia         8
Superficie       10
Precio_venta     16
Fecha_venta       0
Vendedor          1
Edad_vendedor    56
dtype: int64

In [159]:
df_coerce.dtypes

Tipo                     object
Operación                object
Provincia                object
Superficie              float64
Precio_venta            float64
Fecha_venta      datetime64[ns]
Vendedor                 object
Edad_vendedor           float64
dtype: object

### Cambiamos signos especiales usando astype

Es el comportamiento que utiliza to_numeric si no puede convertir un valor, raise hará una excepción, no cambia los valores no numéricos.

In [160]:
# Hacemos una copia para no modificar el dataframe original
df_raise = df.copy()

# Usando pd.to_numeric, se cambian los errores a nulos, nos vamos a encontrar con errores
df_raise['Superficie'] = pd.to_numeric(df_raise['Superficie'], errors= 'raise')
df_raise['Precio_venta'] = pd.to_numeric(df_raise['Precio_venta'], errors= 'raise')
df_raise['Edad_vendedor'] = pd.to_numeric(df_raise['Edad_vendedor'], errors= 'raise')

##Si usaramos ignore, simplemente ignora la operación si encuentra un valor invalido

ValueError: Unable to parse string " " at position 72

## Cambiar inconsistencia de variables

Por ejemplo, no puede exitir un tipo de vivienda que se llame pisar, por lo que tenemos que inferirir que se trata de piso, así que cambiaremos el nombre por piso

In [ ]:
#Corregimos la variable usando replace
#Columna Tipo
df_coerce['Tipo']=df_coerce['Tipo'].replace('Pisar','Piso')

#Columna Operacion
df_coerce['Operación']=df_coerce['Operación'].replace(' ',0)

#Columna Superficie
df_coerce['Superficie']=df_coerce['Superficie'].replace(' ',0)

#Columna Precio_venta
df_coerce['Precio_venta']=df_coerce['Precio_venta'].replace(' ',0)

#Columna Edad_vendedor
df_coerce['Edad_vendedor']=df_coerce['Edad_vendedor'].replace('Pisar','Piso')


In [ ]:
#Utilizamos de nuevo value_counts() para ver el nuevo conteo
df_coerce = df_coerce['Tipo'].value_counts()


## Eliminamos outliers y valores vacíos

### Valores vacíos

En este ejercicio vamos a observar que sucede si eliminamos todos los valores vacíos

In [ ]:
df.isnull().sum()

In [ ]:
#Eliminamos los valores NaNs de la tabla
df = df.dropna(axis=0)
df

In [ ]:
#Reiniciamos el Index
df.dropna(axis=0).reset_index().tail(3)


In [ ]:
#Identificamos vacíos usando isna
df.isna().any()

In [ ]:
#Identificamos vacíos usando isnull
df.isnull().sum()

## Imputación de datos

Reemplazar datos con erroes o faltantes, el metodo depende de la cantidad de datos faltantes.

## M2 act 2
## Alvaro José Bolaños Vega - A00838866

## Resumen notebook

- Leemos el archivo xlsx y lo ponemos en un data frame con pandas
- Reseteamos index, añadiendo una columna index dentro del data frame
- Cambiar el nombre de index a ID
- Se inicia un análisis exploratorio. Se ve la forma del data frame y se revisan los primeros y últimos datos
- Se revisan los tipos de datos, muchos de ellos son object cuando podrían ser números o strings simples.
- La columna cuartos tiene muchos datos nulos.
- Se hace un análisis estadístico descriptivo básico
- Se cuenten y ven porcentajes de nulos por columna, cuartos y vendedores tienen muchos nulos
- Se identifican los tipo de Tipo y cuantos hay por cada uno
- Analizando el resto de columnas se ven errores en los nombres(ortografía y símbolos) de los tipos y varios valores con tipo espacio.
- En la columna edad del vendedor se ven errores en los datos como edades no posibles
- PreProcesamiento de datos
- Se hacen arreglos simples cómo renombrar columnas y quitar columnas innecesarias o con muchos errores
- Se eliminan valores no válidos, se renombran datos con errores identificables, se modifican tipos de datos a string y numérico
- En el nuevo df ya no hay errores.
